### PR TITLE
Fix of crash

### PR DESCRIPTION
**Pull request by santucco**
Backend device can have bigger size of buffer than a frontend device. In that case copying of the frontend buffer is made by lines.

### CODE REVIEW

**Comment by andr2000** [comment]
Thank you for the patches, could you please provide more description in the commit message of what the problem was, how it is solved.

File: src/displayBackend/drm/Dumb.cpp [hunk 1]
```diff
@@ -145,8 +148,17 @@ void DumbDrm::copy()
 	}
 
 	DLOG(mLog, DEBUG) << "Copy dumb, handle: " << mBufDrmHandle;
-
-	memcpy(mBuffer, mGnttabBuffer->get(), mSize);
+	if(mStride == mFrontStride)
```
Comment (andr2000) [inline src/displayBackend/drm/Dumb.cpp:151]: space after if

File: src/displayBackend/drm/Dumb.cpp [hunk 2]
```diff
@@ -145,8 +148,17 @@ void DumbDrm::copy()
...
+	{
+		memcpy(mBuffer, mGnttabBuffer->get(), mSize);
+		return;
+	}
+	auto src = reinterpret_cast<unsigned char*>(mGnttabBuffer->get());
+	auto dst = reinterpret_cast<unsigned char*>(mBuffer);
+	for(unsigned int i = 0; i < mHeight; i++)
```
Comment (andr2000) [inline src/displayBackend/drm/Dumb.cpp:158]: space after for

File: src/displayBackend/drm/Dumb.cpp (hunk 1 above)
Comment (santucco) [reply src/displayBackend/drm/Dumb.cpp]: fixed

File: src/displayBackend/drm/Dumb.cpp (hunk 2 above)
Comment (santucco) [reply src/displayBackend/drm/Dumb.cpp]: fixed

**Comment by andr2000** [comment]
> It causes a crash during an attempt to read from the frontend buffer more bytes than it contains.
> In the case if a stride of the frontend buffer is not equal to the stride of the backend buffer every > line from the frontend buffer is copied in the begin of the corresponding line of the backend buffer.

I think the proper fix would be to negotiate the stride, so we do not guess, but know for sure what it is

**Comment by santucco** [comment]
The stride is a driver specific. Frontend and backend can negotiate about the stride, but I do not see a way to negotiate with the driver.
In the case it seems it is a feature of the i915 driver. According to the documentation a stride(pitch) should be aligned up to 32bit or 64bit, but the driver aligns the pitch up to 64 bytes.
https://elixir.bootlin.com/linux/latest/source/drivers/gpu/drm/i915/i915_gem.c#L261
With zcopy it will be more dramatic, there is not possibility to write to drm buffer line by line.
Maybe it makes a sense to check  a frontend's  stride and  a backend's stride and throw an exception if its are not the same, then make corresponding modifications in a frontend.
I hope I have got you right.

File: src/displayBackend/drm/Dumb.cpp [hunk 3]
```diff
@@ -142,20 +142,20 @@ DumbDrm::~DumbDrm()
 
 void DumbDrm::copy()
 {
-	if(!mGnttabBuffer)
+	if (!mGnttabBuffer)
```
Comment (andr2000) [inline src/displayBackend/drm/Dumb.cpp:145]: Please do not touch this line, this is not yours and should be fixed by us ;)

File: src/displayBackend/drm/Dumb.cpp [hunk 4]
```diff
@@ -145,8 +148,17 @@ void DumbDrm::copy()
...
 
 	DLOG(mLog, DEBUG) << "Copy dumb, handle: " << mBufDrmHandle;
-
-	memcpy(mBuffer, mGnttabBuffer->get(), mSize);
+	if(mStride == mFrontStride)
+	{
+		memcpy(mBuffer, mGnttabBuffer->get(), mSize);
+		return;
+	}
+	auto src = reinterpret_cast<unsigned char*>(mGnttabBuffer->get());
```
Comment (andr2000) [inline src/displayBackend/drm/Dumb.cpp:156]: please cast to uint8_t

File: src/displayBackend/drm/Dumb.cpp [hunk 5]
```diff
@@ -145,8 +148,17 @@ void DumbDrm::copy()
...
+	auto dst = reinterpret_cast<unsigned char*>(mBuffer);
```
Comment (andr2000) [inline src/displayBackend/drm/Dumb.cpp:157]: ditto

File: src/displayBackend/drm/Dumb.cpp (hunk 2 above)
Comment (andr2000) [inline src/displayBackend/drm/Dumb.cpp]: use int

**Comment by andr2000** [comment]
Ok, zcopy is always problematic ;)

File: src/displayBackend/drm/Dumb.cpp [hunk 6]
```diff
@@ -51,6 +51,7 @@ DumbBase::DumbBase(int drmFd, uint32_t width, uint32_t height) :
 	mDrmFd(drmFd),
 	mBufDrmHandle(0),
 	mStride(0),
```
Comment (andr2000) [inline src/displayBackend/drm/Dumb.cpp:53]: So, we should probably s/mStride/sBackStride with a dedicated patch then?

**Comment by andr2000** [comment]
Please also put a note in the commit message that this commit only touches buffers without zero-copy

File: src/displayBackend/drm/Dumb.cpp (hunk 2 above)
Comment (santucco) [reply src/displayBackend/drm/Dumb.cpp]: mHeight is uint32_t, will be warning about a comparison between signed and unsigned
I think I should  use uint32_t

Comment (andr2000) [reply src/displayBackend/drm/Dumb.cpp]: makes sense, then just leave it as is

**Comment by andr2000** [comment]
LGTM:
1. Please make mStride renaming the very first patch
2. And then the fix
3. Please add Signed-off-by
4. Rebase
Thank you
You can add my `Reviewed-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>` if the above done

**Comment by andr2000** [comment]
And also could you please more details in the commit messages?

**Comment by santucco** [comment]
It seems there are some problems with the pull request, I think it is because of my efforts.
It is even better. I can make an other pull request, where I have support a possibility for a frontend to make a buffer with a corresponding size. I have considered all of your requirements, but I have not added of your review tag. Let's close this pull request. What do you think about it?

**Comment by andr2000** [comment]
whatever works best for you

**Pull request closed without merging**